### PR TITLE
Use Qr factorization to get orthogonal matrix in orthogonal init

### PR DESCRIPTION
**Pull request by GBLin5566**
Use qr factorization (instead of svd) to get better performance.

- QR is O(mn^2 - n^3 / 3)

- SVD is O(mn^2 + n^3)

### CODE REVIEW

File: torch/nn/init.py [hunk 1]
```diff
@@ -309,13 +309,19 @@ def orthogonal(tensor, gain=1):
...
-    if u.is_same_size(flattened):
-        tensor.view_as(u).copy_(u)
-    else:
-        tensor.view_as(v.t()).copy_(v.t())
-
+    # Compute the qr factorization
+    q, r = torch.qr(flattened)
+    # Make Q uniform according to https://arxiv.org/pdf/math-ph/0609050.pdf
+    squae_len = cols if rows > cols else rows
+    d = torch.diag(r[:squae_len, :squae_len], 0)
```
Comment (apaszke) [inline torch/nn/init.py]: I don't think you need to make the matrix square beforehand. `torch.diag` works with rectangular matrices as well.

File: torch/nn/init.py [hunk 2]
```diff
@@ -309,13 +309,19 @@ def orthogonal(tensor, gain=1):
...
+    ph = d / torch.abs(d)
```
Comment (apaszke) [inline torch/nn/init.py]: It seems that it's equivalent to `d.sign()`.

File: torch/nn/init.py [hunk 3]
```diff
@@ -309,13 +309,19 @@ def orthogonal(tensor, gain=1):
...
+    q *= ph.expand_as(q)
+    # Pad zeros to Q (if rows smaller than cols)
+    if rows < cols:
+        padding = torch.zeros(rows, cols - rows)
+        q = torch.cat([q, padding], 1)
```
Comment (apaszke) [inline torch/nn/init.py]: This only works if parameters are located on the CPU (otherwise, you'll be doing a `cat` of a GPU and CPU tensor).

**Review by apaszke — COMMENTED** [review]
Is the performance difference really noticeable? Did you benchmark it?

**Comment by GBLin5566** [comment]
@apaszke thanks for commenting the code, fix it already.
About the benchmark, I make a simple performance test on my (poor) mac air.
I fix the seed and make a huge matrix init.
Result:
- QR factorization
('orthogonal ', (1000L, 10000L), ' done in ', 3.1053740978240967, 'seconds')
('orthogonal ', (10000L, 1000L), ' done in ', 2.580883026123047, 'seconds')
('orthogonal ', (3000L, 3000L), ' done in ', 4.194360017776489, 'seconds')
- SVD
('orthogonal ', (1000L, 10000L), ' done in ', 10.98503303527832, 'seconds')
('orthogonal ', (10000L, 1000L), ' done in ', 8.293554067611694, 'seconds')
('orthogonal ', (3000L, 3000L), ' done in ', 210.0512809753418, 'seconds')

As the result. when n is huge (3000 in this case), the running time of SVD will become very scary and QR factorization will be fine. And I think it can be proved in theory too.
(comparing - n^3 / 3 and n^3 in QR and SVD)

**Comment by GBLin5566** [comment]
@apaszke hi I check the ci error message and I got a error when ci test the dataloaders.
Here's the error message,
`FAIL: test_sequential_batch (__main__.TestDictDataLoader)`
But I didn't change anything in dataloader and it seems dataloader did not call orthogonal init. too. I don't understand why it happened. Is this a bug in master's code?

**Review by soumith — APPROVED** [review]

**Comment by soumith** [comment]
thank you @GBLin5566
